File: src/commands.rs
use serde_json::Value;
use tauri::{Manager, Runtime};

use crate::PolodbExt;

#[tauri::command]
pub async fn list_databases<R: Runtime>(
    app: tauri::AppHandle<R>,
) -> Result<Vec<String>, crate::Error> {
    app.polodb().list_databases().await
}

#[tauri::command]
pub async fn open_database<R: Runtime>(
    app: tauri::AppHandle<R>,
    key: String,
    path: String,
) -> Result<String, crate::Error> {
    if let Ok(buf) = app.path().parse(path) {
        app.polodb()
            .open_database(key, buf.to_str().unwrap().to_string())
            .await
    } else {
        Err(crate::Error::Io("Invalid path".to_string()))
    }
}

#[tauri::command]
pub async fn close_database<R: Runtime>(
    app: tauri::AppHandle<R>,
    key: String,
) -> Result<String, crate::Error> {
    app.polodb().close_database(key).await
}

#[tauri::command]
pub async fn insert<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    documents: Vec<Value>,
) -> Result<Vec<usize>, crate::Error> {
    app.polodb().insert(database, collection, documents).await
}

#[tauri::command]
pub async fn insert_one<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    document: Value,
) -> Result<Vec<usize>, crate::Error> {
    app.polodb()
        .insert(database, collection, vec![document])
        .await
}

#[tauri::command]
pub async fn find<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    query: Value,
    sort: Option<Value>,
) -> Result<Vec<Value>, crate::Error> {
    match sort {
        Some(sorting) => {
            app.polodb()
                .find_sorted(database, collection, query, sorting)
                .await
        }
        None => app.polodb().find(database, collection, query).await,
    }
}

#[tauri::command]
pub async fn find_all<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    sort: Option<Value>,
) -> Result<Vec<Value>, crate::Error> {
    match sort {
        Some(sorting) => app.polodb().all_sorted(database, collection, sorting).await,
        None => app.polodb().all(database, collection).await,
    }
}

#[tauri::command]
pub async fn find_one<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    query: Value,
) -> Result<Value, crate::Error> {
    app.polodb().find_one(database, collection, query).await
}

#[tauri::command]
pub async fn delete<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    query: Value,
) -> Result<u64, crate::Error> {
    app.polodb().delete(database, collection, query).await
}

#[tauri::command]
pub async fn delete_one<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    query: Value,
) -> Result<u64, crate::Error> {
    app.polodb().delete_one(database, collection, query).await
}

#[tauri::command]
pub async fn delete_all<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String
) -> Result<u64, crate::Error> {
    app.polodb().delete_all(database, collection).await
}

#[tauri::command]
pub async fn update<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    query: Value,
    update: Value,
    upsert: bool
) -> Result<u64, crate::Error> {
    app.polodb().update(database, collection, query, update, upsert).await
}

#[tauri::command]
pub async fn update_one<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    query: Value,
    update: Value,
    upsert: bool
) -> Result<u64, crate::Error> {
    app.polodb().update_one(database, collection, query, update, upsert).await
}

#[tauri::command]
pub async fn update_all<R: Runtime>(
    app: tauri::AppHandle<R>,
    database: String,
    collection: String,
    update: Value,
    upsert: bool
) -> Result<u64, crate::Error> {
    app.polodb().update_all(database, collection, update, upsert).await
}

File: guest-js/commands.ts
import { invoke, InvokeArgs } from "@tauri-apps/api/core";
import { Result } from "./types";

async function exec<R = any>(
    command: string,
    args?: InvokeArgs
): Promise<Result<R>> {
    try {
        return {
            success: true,
            data: await invoke<R>(`plugin:polodb|${command}`, args),
        };
    } catch (e) {
        try {
            return {
                success: false,
                error: Object.keys(e as object)[0],
                context: (e as { [key: string]: any })[
                    Object.keys(e as object)[0]
                ],
            };
        } catch (u) {
            return {
                success: false,
                error: "unknown",
                context: u,
            };
        }
    }
}

export async function list_databases(): Promise<Result<string[]>> {
    return await exec<string[]>("list_databases");
}

export async function list_collections(
    database: string
): Promise<Result<string[]>> {
    return await exec<string[]>("list_collections", { database });
}

export async function open_database(
    key: string,
    path: string
): Promise<Result<string>> {
    return await exec<string>("open_database", { key, path });
}

export async function close_database(key: string): Promise<Result<string>> {
    return await exec<string>("close_database", { key });
}

export async function insert<T extends object = any>(
    database: string,
    collection: string,
    documents: T[]
): Promise<Result<number>> {
    return await exec<number>("insert", { database, collection, documents });
}

export async function insert_one<T extends object = any>(
    database: string,
    collection: string,
    document: T
): Promise<Result<number>> {
    return await exec<number>("insert_one", { database, collection, document });
}

export async function find_many<
    Document extends object = any,
    Query extends object = any,
    Sorting extends object = any
>(
    database: string,
    collection: string,
    query: Query,
    sorting?: Sorting | null
): Promise<Result<Document[]>> {
    return await exec<Document[]>("find", {
        database,
        collection,
        query,
        sorting: sorting ?? null,
    });
}

export async function find_all<
    Document extends object = any,
    Sorting extends object = any
>(
    database: string,
    collection: string,
    sorting?: Sorting | null
): Promise<Result<Document[]>> {
    return await exec<Document[]>("find_all", {
        database,
        collection,
        sorting: sorting ?? null,
    });
}

export async function find_one<
    Document extends object = any,
    Query extends object = any
>(
    database: string,
    collection: string,
    query: Query
): Promise<Result<Document>> {
    return await exec<Document>("find_one", {
        database,
        collection,
        query,
    });
}

export async function delete_many<Query extends object = any>(
    database: string,
    collection: string,
    query: Query
): Promise<Result<number>> {
    return await exec<number>("delete", {
        database,
        collection,
        query,
    });
}

export async function delete_all(
    database: string,
    collection: string
): Promise<Result<number>> {
    return await exec<number>("delete_all", {
        database,
        collection,
    });
}

export async function delete_one<Query extends object = any>(
    database: string,
    collection: string,
    query: Query
): Promise<Result<number>> {
    return await exec<number>("delete_one", {
        database,
        collection,
        query,
    });
}

export async function update_many<
    Query extends object = any,
    Update extends object = any
>(
    database: string,
    collection: string,
    query: Query,
    update: Update,
    upsert?: boolean
): Promise<Result<number>> {
    return await exec<number>("update", {
        database,
        collection,
        query,
        update,
        upsert: upsert ?? false,
    });
}

export async function update_all<Update extends object = any>(
    database: string,
    collection: string,
    update: Update,
    upsert?: boolean
): Promise<Result<number>> {
    return await exec<number>("update_all", {
        database,
        collection,
        update,
        upsert: upsert ?? false,
    });
}

export async function update_one<
    Query extends object = any,
    Update extends object = any
>(
    database: string,
    collection: string,
    query: Query,
    update: Update,
    upsert?: boolean
): Promise<Result<number>> {
    return await exec<number>("update_one", {
        database,
        collection,
        query,
        update,
        upsert: upsert ?? false,
    });
}
